op-chain-ops/cmd/op-simulate/main.go

```diff
@@ -0,0 +1,304 @@
+	return ethclient.NewClient(cl).TransactionReceipt(ctx, txHash)
+}
+
+func fetchHeader(ctx context.Context, cl *rpc.Client, blockHash common.Hash) (*types.Header, error) {
+	return ethclient.NewClient(cl).HeaderByHash(ctx, blockHash)
+}
+
+type DumpAccount struct {
+	Balance hexutil.Big                 `json:"balance"`
+	Nonce   uint64                      `json:"nonce"`
+	Code    hexutil.Bytes               `json:"code,omitempty"`
+	Storage map[common.Hash]common.Hash `json:"storage,omitempty"`
+}
+
+func readDump(prestatePath string) (map[common.Address]DumpAccount, error) {
+	f, err := os.Open(prestatePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load state data: %w", err)
+	}
+	var out map[common.Address]DumpAccount
+	if err := json.NewDecoder(f).Decode(&out); err != nil {
+		return nil, fmt.Errorf("failed to decode state data: %w", err)
+	}
+	return out, nil
+}
+
+type simChainContext struct {
+	eng  consensus.Engine
+	head *types.Header
+}
+
+func (d *simChainContext) Engine() consensus.Engine {
+	return d.eng
+}
+
+func (d *simChainContext) GetHeader(h common.Hash, n uint64) *types.Header {
+	if n == d.head.Number.Uint64() {
+		return d.head
+	}
+	panic(fmt.Errorf("header retrieval not supported, cannot fetch %s %d", h, n))
+}
+
+func simulate(ctx context.Context, logger log.Logger, conf *params.ChainConfig,
+	prestatePath string, tx *types.Transaction, header *types.Header, doProfile bool) error {
+	memDB := rawdb.NewMemoryDatabase()
+	stateDB := gstate.NewDatabase(memDB)
+	state, err := gstate.New(types.EmptyRootHash, stateDB, nil)
+	if err != nil {
+		return fmt.Errorf("failed to create in-memory state: %w", err)
+	}
+	dump, err := readDump(prestatePath)
+	if err != nil {
+		return fmt.Errorf("failed to read prestate: %w", err)
+	}
+	for addr, acc := range dump {
+		state.CreateAccount(addr)
+		state.SetBalance(addr, new(big.Int).Set((*big.Int)(&acc.Balance)))
+		state.SetNonce(addr, acc.Nonce)
+		state.SetCode(addr, acc.Code)
+		state.SetStorage(addr, acc.Storage)
+	}
+
+	// load prestate data into memory db state
+	_, err = state.Commit(header.Number.Uint64()-1, true)
+	if err != nil {
+		return fmt.Errorf("failed to write state data to underlying DB: %w", err)
+	}
+
+	rules := conf.Rules(header.Number, true, header.Time)
+	signer := types.MakeSigner(conf, header.Number, header.Time)
+	sender, err := signer.Sender(tx)
+	if err != nil {
+		return fmt.Errorf("failed to get tx sender: %w", err)
+	}
+	// prepare the state
+	precompiles := vm.ActivePrecompiles(rules)
+	state.Prepare(rules, sender, header.Coinbase, tx.To(), precompiles, tx.AccessList())
+	state.SetTxContext(tx.Hash(), 0)
+
+	cCtx := &simChainContext{eng: beacon.NewFaker(), head: header}
+	gp := core.GasPool(tx.Gas())
+	usedGas := uint64(0)
+	vmConfig := vm.Config{}
+
+	if doProfile {
+		prof := profile.Start(profile.NoShutdownHook, profile.ProfilePath("."), profile.CPUProfile)
```

Comment:
It could be worth adding a trace profile to get more info about where exactly we spend the time.

Reply:
Can you elaborate? You mean a different pprof type? We can add another CLI flag to customize it more

Reply:
yea, more profile types